File: 2018 zimowy/WawaJava2/src/san/pp/Heron.java
package san.pp;

public class Heron {

  public static void main(String[] args) {
    System.out.println(loop(25, 1));
  }

  private static double loop(double x, double y) {
    if (goodEnough(y, x)) {
      return y;
    }
    return loop(x, improved(y, x));
  }

  private static double improved(double y, double x) {
    return (y + x / y) / 2;
  }

  private static boolean goodEnough(double y, double x) {
    return abs(square(y) - x) < 0.000001;
  }

  private static double square(double x) {
    return x * x;
  }

  private static double abs(double x) {
    return x < 0 ? -x : x;
  }
}

File: 2020 letni (Spring)/Clojure (PF)/fp/src/fp/heron.clj
(ns fp.heron)

;; Zadanie. Napisz procedurę obliczającą wartość pierwiastka kwadratowego z liczby x

;; Problem sformułowany deklaratywnie:
;; sqrt(x) dla x >= 0
;; to jest takie y >= 0, że y*y = x

;; Wiedza o rozwiązaniu problemu (imperatwna):
;; 1. Zgadnij (przyjmij początkowe) y
;; 2. Ulepsz y
;; 3. Ulepszaj y tak długo, aż y stanie się wystarczająco dobre

;; def heronSqrt(y, x):
;;   if goodEnough(y, x):
;;     return y
;;   else
;;     return heronSqrt(improve(y, x), x)

(defn square
  [x]
  (* x x))

;; (if <predicate>
;;   <consequent>
;;   <alternative>)

(defn abs
  [x]
  (if (< x 0)
    (- x)
    x))

(defn good-enough?
  [y x]
  (< (abs (- x (square y))) 0.00001))

(defn avg
  [x y]
  (/ (+ x y) 2))

(defn improve
  [y x]
  (avg y (/ x y)))

(defn heron-sqrt
  [y x]
  (if (good-enough? y x)
    y

    (heron-sqrt (improve y x) x)))

(println (double (heron-sqrt 1 25)))
